### PR TITLE
DD-10 update jira db schedule task

### PR DESCRIPTION
**Pull request by ghost**
- add an api endpoint for syncing our db with jira data
- add a new key in the appconfig with the location of the powershell script which will invoke the url (api endpoint)
- add methods for adding/updating/deleting the schedule task
- refactor TaskSchedulerService to avoid duplicating code
- add a data service method to get the instance by the key, to avoid rewriting the same code over and over again

### CODE REVIEW

File: Equilobe.DailyReport.SL/ConfigurationService.cs [hunk 1]
```diff
@@ -57,5 +54,10 @@ public string GetBitBucketApiClientUrl()
         {
             return ConfigurationManager.AppSettings["bitBucketApiClientUrl"];
         }
+
+        public string GetJiraDBSyncToolPath()
+        {
+            return ConfigurationManager.AppSettings["jiraDBSyncToolPath"];
```
Comment (Sebyd) [inline Equilobe.DailyReport.SL/ConfigurationService.cs]: Is this the path to the powershell script?

File: Equilobe.DailyReport.Models/Storage/InstalledInstance.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@ public class InstalledInstance : IInstance
         public long Id { get; set; }
         [Required]
         public string BaseUrl { get; set; }
+        public string UniqueKey { get; set; }
```
Comment (Sebyd) [inline Equilobe.DailyReport.Models/Storage/InstalledInstance.cs:13]: I don't think we need the unique key. We already have the id.

File: Equilobe.DailyReport.SL/ReportService.cs [hunk 3]
```diff
@@ -13,16 +14,35 @@ public class ReportService : IReportService
     {
         public IJiraService JiraService { get; set; }
         public IDataService DataService { get; set; }
+        public ITaskSchedulerService TaskSchedulerService { get; set; }
 
         public JiraRequestContext JiraRequestContext { get; set; }
 
         #region IReportService Implementation
+        public SimpleResult TrySyncJiraDB(string instanceUniqueKey)
```
Comment (Sebyd) [inline Equilobe.DailyReport.SL/ReportService.cs]: Why is this method called "TrySyncJiraDB"? It doesn't have a try catch inside. Usually that is the convention.

File: Equilobe.DailyReport.SL/TaskSchedulerService.cs [hunk 4]
```diff
@@ -26,37 +25,56 @@ public bool TryRunReportTask(ProjectContext context)
             }
         }
 
+        public void CreateJiraDBSyncTask(string instanceUniqueKey)
```
Comment (Sebyd) [inline Equilobe.DailyReport.SL/TaskSchedulerService.cs]: Find a better name for this. JiraDb makes you think it is the jira database.

**Review by Sebyd — CHANGES_REQUESTED** [review]

File: Equilobe.DailyReport.SL/ConfigurationService.cs (hunk 1 above)
Comment (ghost) [reply Equilobe.DailyReport.SL/ConfigurationService.cs]: Yes

File: Equilobe.DailyReport.Models/Storage/InstalledInstance.cs (hunk 2 above)
Comment (ghost) [reply Equilobe.DailyReport.Models/Storage/InstalledInstance.cs:13]: I thought that using the id is simple enough to make multiple spam requests. Having an unique key, it's way harder to guess

File: Equilobe.DailyReport.SL/ReportService.cs (hunk 3 above)
Comment (ghost) [reply Equilobe.DailyReport.SL/ReportService.cs]: I had a try-catch and forgot to rename the method after removing it.

File: Equilobe.DailyReport.Models/Storage/InstalledInstance.cs (hunk 2 above)
Comment (ghost) [reply Equilobe.DailyReport.Models/Storage/InstalledInstance.cs:13]: As a basic security measure

File: Equilobe.DailyReport.SL/ConfigurationService.cs [hunk 5]
```diff
@@ -57,5 +54,15 @@ public string GetBitBucketApiClientUrl()
         {
             return ConfigurationManager.AppSettings["bitBucketApiClientUrl"];
         }
+
+        public string GetDashboardDataSyncScriptPath()
+        {
+            return ConfigurationManager.AppSettings["dashboardDataSyncScriptPath"];
+        }
+
+        public string GetPowershell()
```
Comment (Sebyd) [inline Equilobe.DailyReport.SL/ConfigurationService.cs]: What is this for? Don't we have the script from GetDashboardDataSyncScriptPath?

File: Equilobe.DailyReport.SL/TaskSchedulerService.cs [hunk 6]
```diff
@@ -123,6 +155,17 @@ private TaskFolder GetTaskFolder(TaskService taskService)
...
+
+        private string GetSyncTaskArguments(string instanceUniqueKey)
+        {
+            var instance = DataService.GetInstanceByKey(instanceUniqueKey);
+            var apiEndpoint = "\"http://localhost:59489/api/dashboardSync?instanceUniqueKey=" + instanceUniqueKey + "\"";
+            var scriptPath = ConfigurationService.GetDashboardDataSyncScriptPath();
+
+            return scriptPath + " -ApiEndpoint " + apiEndpoint;
+
+            //return instance.BaseUrl + "report/syncDashboardData?instanceUniqueKey=" + instanceUniqueKey;
```
Comment (Sebyd) [inline Equilobe.DailyReport.SL/TaskSchedulerService.cs]: remove this commented line if it is not ued anymroe

File: Equilobe.DailyReport.SL/ConfigurationService.cs (hunk 5 above)
Comment (ghost) [reply Equilobe.DailyReport.SL/ConfigurationService.cs]: I found that I cannot run the script on its own. I have to start a scheduled task with the program Powershell.exe having the arguments the script and its arguments;

**Review by Sebyd — APPROVED** [review]

**Pull request merged**